test/integration/ldap_client_test.go

```diff
@@ -0,0 +1,553 @@
+			username:            "pinny",
+			password:            pinnyPassword,
+			provider:            upstreamldap.New(*providerConfig(func(p *upstreamldap.ProviderConfig) { p.UserSearch.Base = "ou=groups,dc=pinniped,dc=dev" })),
+			wantUnauthenticated: true,
+		},
+		{
+			name:                "when there is no username specified",
+			username:            "",
+			password:            pinnyPassword,
+			provider:            upstreamldap.New(*providerConfig(nil)),
+			wantUnauthenticated: true,
+		},
+		{
+			name:      "when there is no password specified",
+			username:  "pinny",
+			password:  "",
+			provider:  upstreamldap.New(*providerConfig(nil)),
+			wantError: `error binding for user "pinny" using provided password against DN "cn=pinny,ou=users,dc=pinniped,dc=dev": LDAP Result Code 206 "Empty password not allowed by the client": ldap: empty password not allowed by the client`,
+		},
+		{
+			name:                "when the user has no password in their entry",
+			username:            "olive",
+			password:            "anything",
+			provider:            upstreamldap.New(*providerConfig(nil)),
+			wantUnauthenticated: true,
+		},
+	}
+
+	for _, test := range tests {
+		tt := test
+		t.Run(tt.name, func(t *testing.T) {
+			authResponse, authenticated, err := tt.provider.AuthenticateUser(ctx, tt.username, tt.password)
+
+			switch {
+			case tt.wantError != "":
+				require.EqualError(t, err, tt.wantError)
+				require.False(t, authenticated, "expected the user not to be authenticated, but they were")
+				require.Nil(t, authResponse)
+			case tt.wantUnauthenticated:
+				require.NoError(t, err)
+				require.False(t, authenticated, "expected the user not to be authenticated, but they were")
+				require.Nil(t, authResponse)
+			default:
+				require.NoError(t, err)
+				require.True(t, authenticated, "expected the user to be authenticated, but they were not")
+				require.Equal(t, tt.wantAuthResponse, authResponse)
+			}
+		})
+	}
+}
+
+func TestSimultaneousRequestsOnSingleProvider(t *testing.T) {
+	env := library.IntegrationEnv(t)
+
+	// Note that these tests depend on the values hard-coded in the LDIF file in test/deploy/tools/ldap.yaml.
+	// It requires the test LDAP server from the tools deployment.
+	if len(env.ToolsNamespace) == 0 {
+		t.Skip("Skipping test because it requires the test LDAP server in the tools namespace.")
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	t.Cleanup(func() {
+		cancelFunc() // this will send SIGKILL to the subprocess, just in case
+	})
+
+	ldapHostPort := findRecentlyUnusedLocalhostPorts(t, 1)[0]
+
+	// Expose the the test LDAP server's TLS port on the localhost.
+	startKubectlPortForward(ctx, t, ldapHostPort, "ldaps", "ldap", env.ToolsNamespace)
+
+	provider := upstreamldap.New(*defaultProviderConfig(env, ldapHostPort))
+
+	// Making multiple simultaneous requests on the same upstreamldap.Provider instance should all succeed
+	// without triggering the race detector.
```

Comment:
Do we actually run the race detector on integration tests 🤔

Reply:
Yes, both in CI and on our laptops.